### DISCUSSION TITLE
I wonder if it's possible to choose server-supported ALPN protocols on per-connection basis

### DISCUSSION THREAD

**Discussion by segfault-bilibili** [nodejs/help#5002] [Ideas]
I'm new to NodeJS. Apologies for my lack of knowledge.

Now we already have `SNICallback` which is able to specify certificates based on `servername`. I'm curious whether it's possible to have something similar, like `ALPNCallback` which (asynchronously) choose ALPN?

I think this might be useful if the server acts like a reverse proxy. If the client supports http/2 while the upstream server doesn't, it seems to help.

**Comment by bnoordhuis** [comment]
That's already supported, if I understand you correctly. See the tls documentation for the `ALPNProtocols` option and the `tlsSocket.alpnProtocol` property.

**Reply by segfault-bilibili** [reply]
I was talking about the server side. I meant the server should be able to behave differently on per-connection basis. However (maybe I just overlooked or misunderstood something) I didn't find any way to do this.

`ALPNProtocols` seems to be a global and immutable option for the server. As for `tlsSocket.alpnProtocol`, it seems to be read-only.

**Reply by bnoordhuis** [reply]
It sounds like you want to look at the ClientHello during the handshake and have the server customize its advertised ALPN response based on that?

Node could be made to support that but... why? What exactly would you customize?

**Reply by segfault-bilibili** [reply]
(If I didn't misunderstand the situation) I've seen CDN servers doing that. It replies different ALPN choices according to SNI domain name, like, if the upstream server doesn't suppport HTTP/2, it won't advertise `h2` through ALPN.

**Reply by bnoordhuis** [reply]
Okay, I can see how that would be useful. Node could support that if you know such facts upfront about your upstream.

What wouldn't work is to connect asynchronously to upstream to see if it speaks `h2`, then report that to the client. openssl doesn't support that, it needs to be synchronous.

**Comment by pimterry** [comment]
Just a quick update on this, after some more related discussion (#45056) there's now a PR to implement this `ALPNCallback` idea here: https://github.com/nodejs/node/pull/45190 (but note that it's synchronous only, due to OpenSSL restrictions)
